### DISCUSSION TITLE
How to create UnionType with 3+ sub-types?

### DISCUSSION THREAD

**Discussion by brandonkelly** [rectorphp/rector#7028] [Q&A] [answered]
There are a few places I need to add union type declarations via `AddPropertyTypeDeclarationRector`, which include 3 or more sub-types, for example:

```diff
- public $plugin = null;
+ public PluginInterface|string|null $plugin = null;
```

However `PHPStan\Type\UnionType` is throwing a `ShouldNotHappenException` from its constructor immediately when attempting to instantiate a `UnionType` with more than two sub-types:

```php
use PHPStan\Type\NullType;
use PHPStan\Type\ObjectType;
use PHPStan\Type\StringType;
use PHPStan\Type\UnionType;

$pluginType = new ObjectType('craft\base\PluginInterface');
$stringType = new StringType();
$nullType = new NullType();

// causes ShouldNotHappenException exception:
$unionType = new UnionType([$pluginType, $stringType, $nullType]);
```

Is there a different way I could go about this with Rector? Or does this just need to be fixed in PHPStan first?

**Comment by TomasVotruba** [comment, marked as answer]
At the moment it's limitation of PHPStan `UnionType` object.

We go around it like this:
https://github.com/rectorphp/rector-symfony/blob/91fd3f3882171c6f0c7e60c44e689e8d7d8ad0a4/config/sets/symfony/symfony6/symfony-return-types.php#L56-L63

**Reply by brandonkelly** [reply]
Thanks! Works like a charm.
